types/time.go

```diff
@@ -596,9 +635,347 @@ func splitDateTime(format string) (seps []string, fracStr string) {
+		var sepsLen = len(seps)
+
+		// Date format must start with number.
+		if i == 0 {
+			if !unicode.IsNumber(rune(format[i])) {
+				return nil, realFsp, err
+			}
+		}
+
+		if i > 0 {
+			var isNumberI_1 = unicode.IsNumber(rune(format[i-1]))
+			var isNumberI = unicode.IsNumber(rune(format[i]))
+
+			//if i-1 is not number,i is number,meaning the beginning of one sep
+			if !isNumberI_1 && isNumberI {
+				start = i
+				lastSepChar = format[i-1]
+			}
+
+			//if i - 1 is number,i is not number,meaning the ending of one sep
+			if isNumberI_1 && !isNumberI {
+				var currentPart = format[start:i]
+				currentSepChar = format[i]
+
+				if sepsLen == 0 {
+					if '-' == lastSepChar || '-' == currentSepChar {
+						seps = append(seps, currentPart)
+						totalPartCount++
+					} else {
+
+						var partsWithYear, validLen, err = ParseNoDelimiterHavingYear(currentPart)
+						if err != nil {
+							return nil, realFsp, err
+						}
+
+						for j := 0; j < validLen; j++ {
+							seps = append(seps, partsWithYear[j])
+						}
+						totalPartCount = validLen
+					}
+				} else {
+					if sepsLen == 6 {
+						if '.' == lastSepChar {
+							if selfDecideFsp {
+								if '0' == currentPart[0] {
+									realFsp = len(currentPart)
+									if realFsp > MaxFsp {
+										realFsp = MaxFsp
+									}
+								} else {
+									realFsp = MaxFsp
+								}
+							}
+							seps = append(seps, currentPart)
+						}
+						return seps, realFsp, nil
+					} else {
+						if '-' == lastSepChar || '-' == currentSepChar {
+							seps = append(seps, currentPart)
+							totalPartCount++
+						} else {
+							var partsNoYear, validLen, err = ParseNoDelimiterNoYear(currentPart)
+							if err != nil {
+								return nil, realFsp, err
+							}
+							for j := 0; j < maxPartCount-sepsLen && j < validLen; j++ {
+								seps = append(seps, partsNoYear[j])
+								totalPartCount++
+							}
+						}
+					}
+				}
+
+				if totalPartCount >= maxPartCount {
+					return seps, realFsp, nil
+				}
+			}
+		}
+	}
+
+	var currentPart = format[start:]
+	var sepsLen = len(seps)
+	if sepsLen == 0 {
+		var partsWithYear, validLen, err = ParseNoDelimiterHavingYear(currentPart)
+		if err != nil {
+			return nil, realFsp, err
+		}
+
+		for j := 0; j < validLen; j++ {
+			seps = append(seps, partsWithYear[j])
+		}
+		totalPartCount = validLen
+	} else if 6 == sepsLen {
+		if '.' == lastSepChar {
+			if selfDecideFsp {
+				if '0' == currentPart[0] {
+					realFsp = len(currentPart)
+					if realFsp > MaxFsp {
+						realFsp = MaxFsp
+					}
+				} else {
+					realFsp = MaxFsp
+				}
+			}
+
+			seps = append(seps, currentPart)
+		}
+
+		return seps, realFsp, nil
+	} else {
+		var partsNoYear, validLen, err = ParseNoDelimiterNoYear(currentPart)
+		if err != nil {
+			return nil, realFsp, err
+		}
+		for j := 0; j < maxPartCount-sepsLen && j < validLen; j++ {
+			seps = append(seps, partsNoYear[j])
+			totalPartCount++
+		}
+	}
+
+	return seps, realFsp, nil
```

Comment:
For test case "011011.12.43", the old version splitDateTime can not work because it get the fracStr by splitting with “.”.
For many other new test case,the old version is hard to adapte.
For improving readability,codes following the for loop can be moved into another func named parseLastPart.

Reply:
Readability is really important here for us to maintain the code. @laidahe